internal/pkg/gateway/api.go

```diff
@@ -86,24 +91,90 @@ func (gs *Server) Endorse(ctx context.Context, request *gp.EndorseRequest) (*gp.
+	channel, chaincodeID, hasTransientData, err := getChannelAndChaincodeFromSignedProposal(signedProposal)
 	if err != nil {
 		return nil, status.Errorf(codes.InvalidArgument, "failed to unpack transaction proposal: %s", err)
 	}
 
-	var endorsers []*endorser
-	if len(request.EndorsingOrganizations) > 0 {
-		endorsers, err = gs.registry.endorsersForOrgs(channel, chaincodeID, request.EndorsingOrganizations)
-	} else {
-		endorsers, err = gs.registry.endorsers(channel, chaincodeID)
-	}
+	err = gs.registry.registerChannel(channel)
 	if err != nil {
 		return nil, status.Errorf(codes.Unavailable, "%s", err)
 	}
 
 	ctx, cancel := context.WithTimeout(ctx, gs.options.EndorsementTimeout)
 	defer cancel()
 
+	defaultInterest := &peer.ChaincodeInterest{
+		Chaincodes: []*peer.ChaincodeCall{{
+			Name: chaincodeID,
+		}},
+	}
+
+	var endorsers []*endorser
+	var responses []*peer.ProposalResponse
+	if len(request.EndorsingOrganizations) > 0 {
+		// The client is specifying the endorsing orgs and taking responsibility for ensuring it meets the signature policy
+		endorsers, err = gs.registry.endorsersForOrgs(channel, chaincodeID, request.EndorsingOrganizations)
+		if err != nil {
+			return nil, status.Errorf(codes.Unavailable, "%s", err)
+		}
+	} else {
+		// The client is delegating choice of endorsers to the gateway.
+
+		// 1. Choose an endorser from the gateway's organization
+		var firstEndorser *endorser
+		es, ok := gs.registry.endorsersByOrg(channel, chaincodeID)[gs.registry.localEndorser.mspid]
+		if !ok {
+			// No local org endorsers for this channel/chaincode. If transient data is involved, return error
+			if hasTransientData {
+				return nil, status.Error(codes.FailedPrecondition, "no endorsers found in the gateway's organization; retry specifying endorsing organization(s) to protect transient data")
+			}
+			// Otherwise, just let discovery pick one.
+			endorsers, err = gs.registry.endorsers(channel, defaultInterest, "")
+			if err != nil {
+				return nil, status.Errorf(codes.Unavailable, "%s", err)
+			}
+			firstEndorser = endorsers[0]
+		} else {
+			firstEndorser = es[0].endorser
+		}
+
+		// 2. Process the proposal on this endorser
+		firstResponse, err := firstEndorser.client.ProcessProposal(ctx, signedProposal)
+		if err != nil {
+			return nil, rpcError(codes.Aborted, "failed to endorse transaction", endpointError(firstEndorser, err))
+		}
+		if firstResponse.Response.Status < 200 || firstResponse.Response.Status >= 400 {
+			return nil, rpcError(codes.Aborted, "failed to endorse transaction", endpointError(firstEndorser, fmt.Errorf("error %d, %s", firstResponse.Response.Status, firstResponse.Response.Message)))
+		}
+
+		// 3. Extract ChaincodeInterest and SBE policies
+		// The chaincode interest could be nil for legacy peers and for chaincode functions that don't produce a read-write set
+		interest := firstResponse.Interest
+		if len(interest.GetChaincodes()) == 0 {
+			interest = defaultInterest
+		}
```

Comment:
A clarification question - Will this condition be ever true unless it's a read-only transaction? If no, then, is there any purpose of proceeding from here? 

Also, that brings another thought, what if it's read-only transaction for pvtdata? That would record some entries in the interest which would give a wrong signal for proceeding.

Reply:
This will be true if:
- The Interest structure in the ProposalResponse is nil, which will be the case for legacy peers
- The chaincode invocation doesn't create a RWset.  This happens in one of the integration tests where the chaincode only sets an event